### DISCUSSION TITLE
[Typescript] Query types weirdness with `ResultType`, `DocType` and `RawDocType` generics

### DISCUSSION THREAD

**Discussion by hasezoey** [Automattic/mongoose#14059] [Q&A]
I am currently trying to fix https://github.com/typegoose/typegoose/issues/870, but i got confused because of the 3 types mentioned in the title, for example:

`Model.find({})`'s return type is:
```ts
mongoose.Query<
    (mongoose.Document<
      unknown,
      {},
      {
        username?: string | undefined;
      }
    > & {
      username?: string | undefined;
    } & {
      _id: mongoose.Types.ObjectId;
    })[],
    mongoose.Document<
      unknown,
      {},
      {
        username?: string | undefined;
      }
    > & {
      username?: string | undefined;
    } & {
      _id: mongoose.Types.ObjectId;
    },
    {},
    {
      username?: string | undefined;
    },
    'find'
  >;
```
(basically `Query<Document<>[], Document<>, {}, ActualRawObject, QueryOperation>`)

but doing a additional `.find` (because it is possible and has the closest types to the query helpers) like `Model.find({}).find({})` has actually the wrong typing for the suggestions in `{}` (for the second `.find`), because the second `.find` uses the `DocType` parameter of the `Query<>` type, but that is set to a `Document<>`, which makes it suggest function like `populate` as a value:
![Screenshot_20231107_172639](https://github.com/Automattic/mongoose/assets/10911626/065db888-2277-4524-b6ad-8a2f55519986)

in addition, fixing it in typegoose by replacing the current `mongoose.QueryWithHelpers<Document<>, Document<>>` with `mongoose.QueryWithHelpers<Document<>, RawClassType>` actually breaks query helper functions which return `this.find` (and other functions) because [those functions all return the generic](https://github.com/Automattic/mongoose/blob/1bfcdf6946861d65de2d1548dc91df0585b6bfed/types/query.d.ts#L361-L451) `DocType` in the `Query<>`'s `ResultType` generic slot, instead of the `ResultType` generic. current code snippit:
```ts
// snippit from query.d.ts
find(
  filter: FilterQuery<DocType>,
  projection?: ProjectionType<DocType> | null,
  options?: QueryOptions<DocType> | null
): QueryWithHelpers<Array<DocType>, DocType, THelpers, RawDocType, 'find'>;
```
(notice the `QueryWithHelpers<DocType[], DocType>`)

where as [the `Model`'s functions](https://github.com/Automattic/mongoose/blob/1bfcdf6946861d65de2d1548dc91df0585b6bfed/types/models.d.ts#L515-L735) return a `ResultType` in the `ResultType` generic slot, snippit:
```ts
// snippit from model.d.ts
find<ResultDoc = THydratedDocumentType>(
  filter: FilterQuery<TRawDocType>,
  projection?: ProjectionType<TRawDocType> | null | undefined,
  options?: QueryOptions<TRawDocType> | null | undefined
): QueryWithHelpers<Array<ResultDoc>, ResultDoc, TQueryHelpers, TRawDocType, 'find'>;
```
(notice the `QueryWithHelpers<ResultDoc[], ResultDoc>`)

---

so my question is: am i doing things wrong, or are the mongoose types wrong?

mongoose 1bfcdf6946861d65de2d1548dc91df0585b6bfed and 8.0.0 (even in 7.6.4)

**Comment by vkarpov15** [comment]
At first glance it looks like Mongoose's types may be wrong. Maybe replace this:

```ts
// snippit from query.d.ts
find(
  filter: FilterQuery<DocType>,
  projection?: ProjectionType<DocType> | null,
  options?: QueryOptions<DocType> | null
): QueryWithHelpers<Array<DocType>, DocType, THelpers, RawDocType, 'find'>;
```

with this?

```
// snippit from query.d.ts
find(
  filter: FilterQuery<RawDocType>,
  projection?: ProjectionType<RawDocType> | null,
  options?: QueryOptions<DocType> | null
): QueryWithHelpers<Array<DocType>, DocType, THelpers, RawDocType, 'find'>;
```

You're right that `find()` filter param shouldn't type as a document.

**Reply by hasezoey** [reply]
that fixes it for the `.find` method, other methods will also need to be touched, in addition aside from the testing case i was working with (plain mongoose), in typegoose i also need to explicitly set Query's `RawDocType` generic now, because otherwise it defaults to `DocType`.

and to clarify, why do those methods return `DocType[]` as the `ResultType` instead of using the generic `ResultType` already available, or does `.find` overwrite something at runtime that i missed (like overwriting the previous call to `.find`)?
the `.find` should also likely have a generic itself to match the definition in `model.d.ts` (`find<ResultDoc = THydratedDocumentType>`), right?
